[sw, rom_ext_signer] Add signing and image update functionality

silvestrst opened the pull request:
# PR Description

`[sw, rom_ext_signer] Add temporary development files`
`[sw, rom_ext_signer] Move configuration file and use dev files`
Adds some binaries for development, debug and testing (expected to be removed eventually).

`[sw, rom_ext_signer] Add signing and remaining image update API`
**Apologies for not splitting this further - I was mixing different changes, and by the end it was too painful to split it out.**

Please read the commit message for this change. It incorporates signing, additional image update API, and includes some minor refactoring.

Aside from some dummy implementations due to uncertainty of how some of the fields or values have to be handled, this is very close to be feature complete.

`e` and `n` extraction is something that we are planning on adding to Mundane. If required a short term solution could be passing these values into the signer similar how we do with `usage_constraints`.

# How to debug

NOTE:
You would need to enable the build in `sw/host/rom_ext_image_tools/signer/meson.build`:
```
...
build_by_default: true,
...
build_by_default: true,
```
Then

`cd $OPENTITAN_TOP`

```
// Run the signer
./build-out/sw/host/rom_ext_image_tools/signer/rom_ext_signer.elf \
    sw/host/rom_ext_image_tools/signer/dev/config.hjson

// Compare before and after
diff <(xxd sw/host/rom_ext_image_tools/signer/dev/rom_ext_blank_image.bin) \
    <(xxd sw/host/rom_ext_image_tools/signer/dev/new_rom_ext_blank_image.bin)
```

mcy commented on sw/host/rom_ext_image_tools/signer/config/src/parser.rs:
```diff
@@ -12,15 +12,16 @@ use serde_hjson::Value;
-    pub image_length: String,
     pub image_version: String,
-    pub image_timestamp: String,
+    pub extension0_offset: String,
     pub extension0_checksum: String,
+    pub extension1_offset: String,
     pub extension1_checksum: String,
+    pub extension2_offset: String,
     pub extension2_checksum: String,
+    pub extension3_offset: String,
     pub extension3_checksum: String,
```
Hmm, do you think this aught to be an array? E.g. `[Extension; 4]`?

mcy commented on sw/host/rom_ext_image_tools/signer/config/src/parser.rs:
```diff
@@ -29,15 +30,13 @@ pub struct ParsedConfig {
 pub struct InputFiles {
     pub image_path: String,
     pub private_key_der_path: String,
+    pub usage_constraints_path: String,
+    pub system_state_value_path: String,
```
Should these be PathBufs instead?

mcy commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -31,51 +38,96 @@ impl RawImage {
+    /// like signature key public exponent and modulus, are obtained at
+    /// run-time.
+    pub fn update_static_fields(&mut self, config: &ParsedConfig) {
+        // TODO: checks to make sure that the config values (strings) are not
+        //       bigger than the actual field size.
```
Can you elaborate on what this would mean?

mcy commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -91,6 +143,53 @@ impl RawImage {
     }
+
+    /// Updates ROM_EXT manifest usage constraints field.
+    fn update_usage_constraints_field(&mut self, dir: &str) {
```
Why doesn't this take `&impl AsRef<Path>`?

mcy commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -91,6 +143,53 @@ impl RawImage {
+    ///
+    /// Generic function, is used by the specific field update counterparts.
+    fn update_field(&mut self, field_data: &[u8], field_offset: u32) {
+        let image = &mut self.data;
```
What does this binding achieve? I think you can just inline it.

mcy commented on sw/host/rom_ext_image_tools/signer/src/main.rs:
```diff
@@ -22,9 +31,86 @@ fn main() {
+        RsaSignature::<B3072, RsaPkcs1v15, Sha256>::sign(&private_key, &message_to_sign)
+            .expect("Failed to sign!");
+
+    // Update signature field.
+    image.update_signature_field(signature.bytes());
+
+    // Write image to disk.
```
These comments say what, not why. Consider removing them?

mcy commented on sw/host/rom_ext_image_tools/signer/src/main.rs:
```diff
@@ -22,9 +31,86 @@ fn main() {
+    // Update manifest fields
+    image.update_static_fields(&config);
+
+    let exponent = &signature_key_public_exponent_le();
```
Again, what's the point of these single-use bindings?

mcy commented on sw/host/rom_ext_image_tools/signer/src/main.rs:
```diff
@@ -22,9 +31,86 @@ fn main() {
+///
+/// Eventually this value will be obtained from the private key.
+fn signature_key_public_exponent_le() -> Vec<u8> {
+    let dummy: Vec<u8> = vec![0xA5; 1];
```
why isn't the body of this function just `vec![0xa5]`?

mcy commented on sw/host/rom_ext_image_tools/signer/src/main.rs:
```diff
@@ -22,9 +31,86 @@ fn main() {
+///
+/// Eventually this value will be obtained from the private key.
+fn signature_key_modulus_le() -> Vec<u8> {
+    let dummy: Vec<u8> = vec![0xA5; 384];
```
Same here. Don't introduce single-use variables that might hurt RVO.

mcy reviewed (commented):
Quick first pass.

As a general rule, avoid single-use variables unless they're necessary to appease the borrow checker. They (can) hurt optimization and introduce unnecessary indirection for readers.

silvestrst commented:
@mcy thank you for your review. I have addressed most of the comments.

> As a general rule, avoid single-use variables unless they're necessary to appease the borrow checker. They (can) hurt optimization and introduce unnecessary indirection for readers.

I think single use variables improve readability in some cases. Saying that, I have changed most of the places where you have pointed it out.

On the optimization subject - I don't think it is a big deal for a host-side tool like this that runs under a second anyway. Moreover, I don't think it is always the best practice to tailor high level code based on some low level compiler concepts (like this host side tool).

silvestrst reviewed (commented):
@mcy thank you for your review. I have addressed the comments.

mcy commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -31,51 +38,97 @@ impl RawImage {
-            &config.extension0_checksum,
+        update(
+            &config.extensions[0].checksum,
             manifest::ROM_EXT_EXTENSION0_CHECKSUM_OFFSET,
         );
-        myclosure(
-            &config.extension1_checksum,
+        update(
+            &config.extensions[1].offset,
+            manifest::ROM_EXT_EXTENSION1_OFFSET_OFFSET,
+        );
+        update(
+            &config.extensions[1].checksum,
             manifest::ROM_EXT_EXTENSION1_CHECKSUM_OFFSET,
         );
-        myclosure(
-            &config.extension2_checksum,
+        update(
+            &config.extensions[2].offset,
+            manifest::ROM_EXT_EXTENSION2_OFFSET_OFFSET,
+        );
+        update(
+            &config.extensions[2].checksum,
             manifest::ROM_EXT_EXTENSION2_CHECKSUM_OFFSET,
         );
-        myclosure(
-            &config.extension3_checksum,
+        update(
+            &config.extensions[3].offset,
+            manifest::ROM_EXT_EXTENSION3_OFFSET_OFFSET,
+        );
+        update(
+            &config.extensions[3].checksum,
             manifest::ROM_EXT_EXTENSION3_CHECKSUM_OFFSET,
         );
```
This kind of duplication is a bit error-prone.

```rs
let offsets = [
  (ROM_EXT_EXTENSION0_OFFSET, ROM_EXT_EXTENSION0_CHECKSUM_OFFSET),
  // ...
];
for (extn_offset, checksum_offset) in offsets.iter().enumerate() {
  self.update_field(&str_to_vec_u8(&config.extensions[i].offset), *extn_offset);
  self.update_field(&str_to_vec_u8(&config.checksums[i].offset), *checksum_offset);
}
```
If you do this, you can get rid of the lambda at the top of this stanza.

mcy commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -31,51 +38,97 @@ impl RawImage {
+
+    /// Updates ROM_EXT manifest signature key public exponent field.
+    pub fn update_exponent_field(&mut self, exponent: &[u8]) {
+        let offset = manifest::ROM_EXT_SIGNATURE_KEY_PUBLIC_EXPONENT_OFFSET;
```
Please inline this. This variable doesn't make the code more readable.

gkelly replied:
It might be in a style guide somewhere, but given that >80c just means work for rustfmt and not you I'm in agreement with @mcy here.

silvestrst replied:
Done.

mcy commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -91,6 +144,50 @@ impl RawImage {
+
+        let dummy: Vec<u8> = vec![0xA5; 16];
+        let offset = manifest::ROM_EXT_PERIPHERAL_LOCKDOWN_INFO_OFFSET;
+        self.update_field(&dummy, offset);
```
This can just be `self.update_field(&[0xa5; 16], offset);`

mcy commented on sw/host/rom_ext_image_tools/signer/src/main.rs:
```diff
@@ -21,10 +30,74 @@ fn main() {
+    image.write_file();
+}
+
+/// Generate a dummy signature key public exponent.
+///
+/// Eventually this value will be obtained from the private key.
+fn signature_key_public_exponent_le() -> Vec<u8> {
+    vec![0xA5; 1]
+}
+
+/// Generate a dummy signature key modulus.
+///
+/// Eventually this value will be obtained from the private key.
+fn signature_key_modulus_le() -> Vec<u8> {
+    vec![0xA5; 384]
+}
```
I feel like it might make more sense to have these be returned together as part of some RsaPrivateKey struct.

silvestrst replied:
This is a temporary dummy, I have already done a change to mundane for propper getters.

mcy commented on sw/host/rom_ext_image_tools/signer/src/main.rs:
```diff
@@ -7,11 +7,20 @@
+use mundane::public::Signature;
+
 fn main() {
     let arg: String = env::args().nth(1).expect("Config path is missing");
```
Prefer args_os, which won't crash on non-UTF-8 input.

mcy reviewed (commented):
Another pass. Also, I noticed this CL doesn't include tests. Can you please add tests so we can be sure the code we're checking in works correctly? I don't think the style we've been doing for DIFs is appropriate for host-side code.

silvestrst commented:
@mcy thanks again for the review!

I just had a quick look, will address issues more likely next week.

Regarding the following:
> Can you please add tests so we can be sure the code we're checking in works correctly?

I absolutely agree with you that this requires tests, however I am trying to push this change forward, and would like to add tests as follow-up change. This is not dissimilar to DIFs, where we split out the effort in header development, implementation (sometimes with tests, sometimes tests follow).

silvestrst reviewed (commented):
@mcy thanks again for the review, I will create an issue tracking outstanding things (testing, etc ...)

silvestrst commented:
See #5232 for the outstanding work.

gkelly commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -31,51 +38,89 @@ impl RawImage {
+    /// of the image, including all the manifest fields that follow the
+    /// signature field.
+    pub fn data_to_sign(&self) -> &[u8] {
+        let offset = manifest::ROM_EXT_SIGNED_AREA_START_OFFSET as usize;
```
Probably same here, but this one actually does something instead of just aliasing the name, so I'm OK with this one here.

silvestrst replied:
Keeping, this one :)

gkelly commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -91,6 +136,49 @@ impl RawImage {
+            .duration_since(UNIX_EPOCH)
+            .expect("Failed to obtain the current time!");
+
+        let offset = manifest::ROM_EXT_IMAGE_TIMESTAMP_OFFSET;
```
In general I'm against single-use variables.

silvestrst replied:
I have fixed all the places in `image.rs` where single use variables simply alias a field or similar.

gkelly reviewed (approved):
Please file an issue for the followup tests, but otherwise looks good modulo some nits about single-use variables.

silvestrst commented on sw/host/rom_ext_image_tools/signer/config/src/parser.rs:
```diff
@@ -12,35 +13,34 @@ use serde_hjson::Value;
 }
 
-/// TODO
+/// ROM_EXT extension.
```
Done.

silvestrst commented on sw/host/rom_ext_image_tools/signer/image/src/image.rs:
```diff
@@ -31,51 +38,89 @@ impl RawImage {
+
+    /// Updates ROM_EXT manifest signature key modulus field.
+    pub fn update_modulus_field(&mut self, modulus: &[u8]) {
+        let offset = manifest::ROM_EXT_SIGNATURE_KEY_MODULUS_OFFSET;
```
Done.

silvestrst commented:
> Please file an issue for the followup tests

Done in #5232
